src/Microsoft.OData.Edm/Csdl/CsdlWriter.cs

```diff
@@ -8,170 +8,169 @@
+            CsdlWriter csdlWriter = new CsdlXmlWriter(model, writer, edmxVersion, target);
+            csdlWriter.WriteCsdl();
 
-        private void WriteEFCsdl()
-        {
-            this.WriteEdmxElement();
-            this.WriteRuntimeElement();
-            this.WriteConceptualModelsElement();
-            this.WriteSchemas();
-            this.EndElement(); // </ConceptualModels>
-            this.EndElement(); // </Runtime>
-            this.EndElement(); // </Edmx>
+            errors = Enumerable.Empty<EdmError>();
+            return true;
         }
 
-        private void WriteEdmxElement()
+        /// <summary>
+        /// Write CSDL output.
+        /// </summary>
+        protected virtual void WriteCsdl()
         {
-            this.writer.WriteStartElement(CsdlConstants.Prefix_Edmx, CsdlConstants.Element_Edmx, this.edmxNamespace);
-            this.writer.WriteAttributeString(CsdlConstants.Attribute_Version, GetVersionString(this.edmxVersion));
+            // nothing here
         }
 
-        private void WriteRuntimeElement()
+        // Gets the string form of the EdmVersion.
+        // Note that Version 4.01 needs two digits of minor version precision.
+        protected static string GetVersionString(Version version)
         {
-            this.writer.WriteStartElement(CsdlConstants.Prefix_Edmx, CsdlConstants.Element_Runtime, this.edmxNamespace);
-        }
+            Debug.Assert(version != null);
 
-        private void WriteConceptualModelsElement()
-        {
-            this.writer.WriteStartElement(CsdlConstants.Prefix_Edmx, CsdlConstants.Element_ConceptualModels, this.edmxNamespace);
-        }
+            if (version == EdmConstants.EdmVersion401)
+            {
+                return EdmConstants.EdmVersion401String;
+            }
 
-        private void WriteReferenceElements()
-        {
-            EdmModelReferenceElementsVisitor visitor = new EdmModelReferenceElementsVisitor(this.model, this.writer, this.edmxVersion);
-            visitor.VisitEdmReferences(this.model);
+            return version.ToString();
         }
 
-        private void WriteDataServicesElement()
+        private static bool VerifyAndGetVersion(IEdmModel model, out Version edmxVersion, out IEnumerable<EdmError> errors)
         {
-            this.writer.WriteStartElement(CsdlConstants.Prefix_Edmx, CsdlConstants.Element_DataServices, this.edmxNamespace);
-        }
+            Debug.Assert(model != null);
 
-        private void WriteSchemas()
-        {
-            // TODO: for referenced model - write alias as is, instead of writing its namespace.
-            EdmModelCsdlSerializationVisitor visitor;
-            Version edmVersion = this.model.GetEdmVersion() ?? EdmConstants.EdmVersionDefault;
-            foreach (EdmSchema schema in this.schemas)
+            edmxVersion = model.GetEdmxVersion();
+
+            errors = model.GetSerializationErrors();
+            if (errors.FirstOrDefault() != null)

```

Comment:
>errors.FirstOrDefault() [](start = 16, length = 23)

Did we ever determine which performed better between errors.FirstOrDefault() and errors.Any()?

Reply:
@mikepizzo Any() is a tiny bit faster than FirstOrDefault() != null, and Any() also doesn't allocate memory for the object.

Reply:
Thanks @bdebaere for the confirmation -- that matches my intuition, but in the past we've sometimes been surprised.